Update HNSW cost estimatation to utilize search and index info

jkatz opened the pull request:
Previously, the cost estimation formula for a HNSW index scan utilized
a methodology that only factored in the entry level for an HNSW scan
and the "m" index parameter, which reflects the number of tuples (or
vectors) to scan at each step of a HNSW graph traversal. While this
would bias the PostgreSQL query planner to choose an HNSW index scan
over other available paths, this could lead to potential suboptimal
index selection, for example, choosing to use a HNSW index instead of
an available B-tree index that has better selectivity.

The number of tuples scanned during HNSW graph traversal is principally
influenced by these factors:

 * The number of tuples stored in the index
 * `m` - the number of tuples that are scanned in each step of the graph
   traversal
 * `hnsw.ef_search` - which influences the total number of steps it
   takes for the scan to converge on the approximated nearest neighbors

Through testing different source models for vectors, we also observed
that the correlation of vectors in mdoels would impact this convergence.
For this first iteration, we've opted to hardcode a constant scaling
factor and set it to `0.55`, though a future commit may turn this into
a configurable parameter.

The high-level formula for estimating the cost of a HNSW index scan is
as such:

```
(entryLevel * m) + (layer0TuplesMax * layer0Selectivity)
```

where

- `(entryLevel * m)` is the lower bound of tuples to scan, as it
accounts for the graph traversal to layer 0 (L0). (L1 and above has an ef=1)
- `layer0TuplesMax` is an estimate of the maximum number of tuples to
scan at L0. This accounts for tuples that may end up being discarded due
to them already being visited. Testing shows that the number of steps
until converge is similar to the value of `hnsw.ef_search`, thus we can
estimate tuples max at `hnsw.ef_search * m * 2`
- `layer0Selectivity` - estimates the percentage of tuples that will
actually be scanned during the index traversal, multipled by the scaling
factor

In addition to the `m` build parameter and `hsnw.ef_search`, costs
estimates can be influenced by standard PostgreSQL costing parameters,
though adjusting those (e.g. `random_page_cost`) should be done with
care.

Co-authored-by: @ankane

jkatz commented:
Note: I've fixed a previous test failure (it was a TODO that's now resolved with the updated cost estimation model :smile:). However, the active failure is interesting. The failing query is:

```sql
EXPLAIN ANALYZE SELECT i FROM tst WHERE v <-> '$query' < 1 ORDER BY v <-> '$query';
```

The dataset size is 10K vectors, and to explicitly state, this is a query with a filter that is ordering the vectors but doesn't have an explicit limit. `hnsw.ef_search` is `40`.

Previously, the costing was such that PostgreSQL estimated the sequential scan was cheaper and opted for that.
With the current patch, PostgreSQL estimates the index scan is cheaper (by almost less than a single cost unit 😨) and opts to use that. This may very well be the fastest scan path, but (untested theory) in streaming mode would this end up using the full index scan, which would certainly be more expensive than a sequential scan?

I have a locally modified version of the patch that borrows similar selectivity logic to IVFFlat (`MIN(our estimate, PostgreSQL's selectivity estimate)`) but PostgreSQL still opted for the sequential scan.

Primarily posting this as a FYI. I'm still pulling a few threads.

ankane commented on src/hnsw.c:
```diff
@@ -119,15 +121,53 @@ hnswcostestimate(PlannerInfo *root, IndexPath *path, double loop_count,
+	 * elimination strategy, whereas HNSW selecitivty is based on scanning the
+	 * index itself.
+	 */
+	costs.indexSelectivity = costs.numIndexTuples / path->indexinfo->tuples;
```
Removing this should fix the test. I don't think this type of selectivity is what the planner expects.

I think we should do the same for IVFFlat ([ivfflat-cost branch](https://github.com/pgvector/pgvector/compare/ivfflat-cost)), which fixes two of the tests.

jkatz replied:
Going to try that; if it works will push up with next commit.

ankane commented on src/hnsw.c:
```diff
@@ -119,15 +121,53 @@ hnswcostestimate(PlannerInfo *root, IndexPath *path, double loop_count,
+	 *
+	 * "HnswSF" is a scaling factor, hardcoded to HnswScanScalingFactor
+	 */
+	totalLayers = floor(log(path->indexinfo->tuples) / log(m));
```
I think it'd be better to keep the previous naming (`entryLevel`) here, as well as `HnswGetMl(m)` since it's really the total number of layers - 1.

We may also want to add `+ 1` to the number of tuples here and below to avoid `log(0)`.

jkatz replied:
@ankane I'm less opinionated on the name; I did try to force `tuples` to 0, but in that case PostgreSQL won't consider the scan. I think in the case where `tuples` is 0, we should max out the costs like we do if there's no ORDER BY quals.

ankane commented on src/hnsw.c:
```diff
@@ -119,15 +121,53 @@ hnswcostestimate(PlannerInfo *root, IndexPath *path, double loop_count,
+	 * The tuple estimator formula is below:
+	 *
+	 * numIndexTuples = (totalLayers * m) +
+	 * 					(layer0TuplesMax * layer0Selectivity * HnswSF)
```
I think `HnswSF` should be included in `layer0Selectivity` to make the `layer0Selectivity` description accurate.

ankane commented on src/hnsw.h:
```diff
@@ -85,6 +85,9 @@
 #define HnswGetMl(m) (1 / log(m))
 
+/* Scaling factor for HNSW index scan estimation */
+#define HnswScanScalingFactor ((double) 0.55)
```
I'd prefer to keep this in `hnswcostestimate`, but if you feel strongly it should be separate, let's use the prefix `HNSW_COST`.

jkatz replied:
I don't feel strongly about that - but - I do feel it should be configurable and just wanted to simplify transition if/when it happens 😛 I will move.

ankane reviewed (commented):
Awesome, thanks for driving this @jkatz! This should significantly improve index selection.

Added some comments inline.

jkatz commented:
@ankane Pushed changes inline with feedback. One discussion point will be the check to push the costs up to the max if tuples is 0.

I'm...OK-ish to go with your suggestion (add `1` to the total number), but arguably if we have no tuples, we shouldn't even bother considering the index scan. Adding 1 is certainly a cheap operation, but it'd be cheaper in that case to not even go through all the operations. OTOH, we then have to check that condition every time, and given it's unlikely we'd encounter that case, it may just be better to "add 1".

ankane commented on src/hnsw.c:
```diff
@@ -99,11 +99,18 @@ hnswcostestimate(PlannerInfo *root, IndexPath *path, double loop_count,
-	if (path->indexorderbys == NULL)
+
+/* Scaling factor for HNSW index scan estimation */
+#define HnswScanScalingFactor ((double) 0.55)
```
Think this was left over.

jkatz replied:
🤦

ankane commented on src/hnsw.c:
```diff
@@ -99,11 +99,18 @@ hnswcostestimate(PlannerInfo *root, IndexPath *path, double loop_count,
+	double		entryLevel;
+	double		layer0TuplesMax;
+	double		layer0Selectivity;
+	double		hnswScalingFactor = 0.55;
```
Let's remove the `hnsw` prefix here and in the comments for simplicity.

ankane commented on src/hnsw.c:
```diff
@@ -119,12 +126,42 @@ hnswcostestimate(PlannerInfo *root, IndexPath *path, double loop_count,
+	 * at L0, accounting for previously visited tuples, multiplied by the
+	 * "hnswScalingFactor" (currently hardcoded).
+	 */
+	entryLevel = floor(log(path->indexinfo->tuples) / log(m));
```
Let's use `* HnswGetMl(m)` instead of `/ log(m)` like the previous code. `ml` will likely never change, but it's really a function of that.

jkatz replied:
@ankane Any reason outside of matching the paper? This way saves a `/` operation.

ankane replied:
It only saves a `1 *` which the compiler likely removes, but I think it's more correct. If `ml` changes, we would need to change this.

jkatz replied:
@ankane Alright, I concede :smile: Pushing it up in a sec.

ankane reviewed (commented):
Great, added a few more comments. I think the `+ 1` approach would be slightly better since the index can fulfill the query (the first path is more designed to prevent something the index can't do).

jkatz commented:
@ankane All feedback should now be addressed.

ankane commented:
Thanks, looks great. Last remaining comment:

> I think the + 1 approach would be slightly better since the index can fulfill the query (the first path is more designed to prevent something the index can't do).

jkatz commented:
@ankane Done.

ankane commented on src/hnsw.c:
```diff
@@ -99,10 +99,13 @@ hnswcostestimate(PlannerInfo *root, IndexPath *path, double loop_count,
 	Relation	index;
 
-	/* Never use index without order */
+	/* Never use index without order or if there are no tuples */
```
This comment needs reverted.

jkatz replied:
🤦 Sorry for the sloppiness - that's due to rushing. Pushed fix.

ankane commented on src/hnsw.c:
```diff
@@ -119,12 +122,42 @@ hnswcostestimate(PlannerInfo *root, IndexPath *path, double loop_count,
+	 * at L0, accounting for previously visited tuples, multiplied by the
+	 * "scalingFactor" (currently hardcoded).
+	 */
+	entryLevel = floor(log(path->indexinfo->tuples) * HnswGetMl(m));
```
Looks like `+ 1` is missing from here.

ankane commented on src/hnsw.c:
```diff
@@ -99,7 +99,10 @@ hnswcostestimate(PlannerInfo *root, IndexPath *path, double loop_count,
 	int			m;
-	int			entryLevel;
+	double		entryLevel;
+	double		layer0TuplesMax;
```
I think the code will be cleaner if an `int` is used for `entryLevel` and `layer0TuplesMax`, but I can do that in a follow-up commit if that's easier.

ankane commented on src/hnsw.c:
```diff
@@ -119,12 +122,42 @@ hnswcostestimate(PlannerInfo *root, IndexPath *path, double loop_count,
+	entryLevel = floor(log(path->indexinfo->tuples) * HnswGetMl(m));
+	layer0TuplesMax = (double) (HnswGetLayerM(m, 0) * hnsw_ef_search);
+	layer0Selectivity =  scalingFactor * 
+		(log(path->indexinfo->tuples + 1) / (log(m) * (1 + log(hnsw_ef_search))));
```
Just noticed there's an extra space here and we can get rid of some parentheses here and below.

ankane reviewed (commented):
Added some more comments, but I can handle the style-related ones in a follow-up commit if that's easier.

jkatz commented:
@ankane Addressed the feedback. I adjusted the parens, but ensuring it aligns to the formula.

![image](https://github.com/user-attachments/assets/a8701046-b873-4be1-9ff7-e0338328d8c4)

ankane commented:
Awesome, thank you! And huge thanks for all of the work that went into deriving the formula!!

jkatz commented:
@ankane 🎉Thanks! Looking forward to describing more of the data...and adding the scaling factor parameter 😉
